### PR TITLE
Improve the Reference model.

### PR DESCRIPTION
**Pull request by frankban**
Also avoid installing sphinx as a system dependency, and fix the docs generation.

### CODE REVIEW

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-bundlelib/21/
Test FAILed.

**Comment by mitechie** [comment]
Manually added sysdeps to the ci server. However we can't run sysdeps automatically via check. jenkins doesn't have access on the system.

**Comment by mitechie** [comment]
:+1: no QA aside from the aptdeps issue.

File: docs/conf.py [hunk 1]
```diff
@@ -63,7 +63,7 @@
 # the built documents.
 #
 # The short X.Y version and the full version.
-version = release = src.get_version()
+version = release = jujubundlelib.get_version()
```
Comment (bac) [inline docs/conf.py]: Is that comment correct about having a short and full version?

File: jujubundlelib/references.py [hunk 2]
```diff
@@ -150,6 +170,13 @@ def is_local(self):
         """Return True if this refers to a local entity, False otherwise."""
         return self.schema == 'local'
 
+    def is_fully_qualified(self):
+        """Report whether this reference is fully qualified.
+
+        A fully qualified reference includes its schema, series and revision.
+        """
+        return self.schema and self.series and (self.revision is not None)
```
Comment (bac) [inline jujubundlelib/references.py]: Maybe use all() here?

File: tox.ini [hunk 3]
```diff
@@ -19,3 +19,8 @@ commands =
 [testenv:lint]
 usedevelop = True
 commands = flake8 --ignore E731 --show-source jujubundlelib
+
+[testenv:docs]
+changedir = docs
+deps = sphinx
+commands = sphinx-build -W -b html -d {envtmpdir}/doctrees . {envtmpdir}/html
```
Comment (bac) [inline tox.ini]: Nice

**Comment by bac** [comment]
Code :+1: with a few suggestions.

So tox implicitly requires sudo now, thus the CI failure?

File: docs/conf.py (hunk 1 above)
Comment (frankban) [inline docs/conf.py]: Yes it is, they are the same most of the times.

File: jujubundlelib/references.py (hunk 2 above)
Comment (frankban) [inline jujubundlelib/references.py]: Not sure, we don't have an iterable already and the current version is quite readable.

**Comment by frankban** [comment]
Not sure about the CI failure, tox does not require sudo and sphinx is installed in a venv.
Thanks for the reviews!
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://localhost:8080/job/juju-bundlelib-merge

**Pull request merged**